dbus: Respect "ptest" in DISTRO_FEATURES

ml-ichiro opened the pull request:
# Purpose of pull request

This PR fixes ptest of dbus package.
After applying this change, ptest of dbus will be installed if ptest is enabled.

Currently ptest of dbus will not be installed even if ptest is enabled.

```
$ . ./repos/poky/oe-init-build-env build
$ bitbake-layers add-layer ../repos/meta-debian/
$ cat << EOS >> conf/local.conf
DISTRO = "deby"
MACHINE = "qemuarm64"
PACKAGE_CLASSES = "package_deb"
DISTRO_FEATURES_append = " ptest"
EXTRA_IMAGE_FEATURES += "ptest-pkgs"
IMAGE_INSTALL_append = " dbus"
EOS
$ bitbake core-image-minimal
...(snip)...
$ runqemu nographic slirp qemuparams="-m 4096"
...(snip)...
$ ptest-runner -l | grep dbus

### "dbus-test" should be shown here, but actually not. ###
```

This is because dbus recipe doesn't respect "ptest" in `DISTRO_FEATURES` variable unlike ptest.bbclass.
Therefore, this PR will fix dbus recipe to make sure it is respected.

# Note

If we build dbus with test (`--enable-tests` in configure), it requires glib-2.0 additionally.
glib-2.0 requires dbus, so it causes circular build-dependency between dbus and glib-2.0.

To avoid this circular build-dependency, test code of dbus is split out as "dbus-test" recipe in current implementation.
As a result, 

* glib-2.0 depends on dbus
* dbus-test depends on glib-2.0.

This workaround makes dbus-test build-able, but the drawback is that it is recognized as "dbus-test" not "dbus" from ptest-runner. I think this is limitation of current implementation.

```
root@qemuarm64:~# ptest-runner -l
Available ptests:
acl     /usr/lib/acl/ptest/run-ptest
busybox /usr/lib/busybox/ptest/run-ptest
dbus-test       /usr/lib/dbus-test/ptest/run-ptest
util-linux      /usr/lib/util-linux/ptest/run-ptest
zlib    /usr/lib/zlib/ptest/run-ptest
```

# Test
## How to test

1. Enable ptest and install dbus package

```
$ . ./repos/poky/oe-init-build-env build
$ bitbake-layers add-layer ../repos/meta-debian/
$ cat << EOS >> conf/local.conf
DISTRO = "deby"
MACHINE = "qemuarm64"
PACKAGE_CLASSES = "package_deb"
DISTRO_FEATURES_append = " ptest"
EXTRA_IMAGE_FEATURES += "ptest-pkgs"
IMAGE_INSTALL_append = " dbus"
EOS
```

2. Build core-image-minimal

```
$ bitbake core-image-minimal
```

3.  Run qemu and run ptest of dbus

```
$ runqemu nographic slirp qemuparams="-m 4096"
...(snip)...
# ptest-runner -l
...(snip)...
# ptest-runner -t 3600 dbus-test
```

Also, I confirmed that SDK builds succeed with the following settings:

* Set `DISTRO=deby` and run `bitbake core-image-minimal -c populate_sdk` with meta-debian and poky
* Set `DISTRO=emlinux` and run `bitbake core-image-minimal-sdk -c populate_sdk` with meta-debian, meta-debian-extended, meta-emlinux and poky

## Test result

```
# ptest-runner -l
Available ptests:
busybox /usr/lib/busybox/ptest/run-ptest
dbus-test       /usr/lib/dbus-test/ptest/run-ptest
util-linux      /usr/lib/util-linux/ptest/run-ptest
zlib    /usr/lib/zlib/ptest/run-ptest
# ptest-runner -t 3600 dbus-test
START: ptest-runner
2024-04-09T00:40
BEGIN: /usr/lib/dbus-test/ptest
[   33.587143] random: test-bus: uninitialized urandom read (12 bytes read)
[   33.593804] random: test-bus: uninitialized urandom read (12 bytes read)
[   33.595391] random: test-bus: uninitialized urandom read (10 bytes read)
dbus-daemon[2037]: dbus-daemon transaction failed (OOM), sending error to sender uid=0 pid=2037 comm="./test/test-bus ./test/data "
dbus-daemon[2037]: dbus-daemon transaction failed (OOM), sending error to sender uid=0 pid=2037 comm="./test/test-bus ./test/data "
dbus-daemon[2037]: dbus-daemon transaction failed (OOM), sending error to sender uid=0 pid=2037 comm="./test/test-bus ./test/data "
...(snip)...
PASS: test/test-syslog
SKIP: test/test-uid-permissions
PASS: test/test-variant
dbus-daemon[3000]: [session uid=0 pid=3000] Reloaded configuration
DURATION: 1094
END: /usr/lib/dbus-test/ptest
2024-04-09T00:58
STOP: ptest-runner
```

[ptest-dbus.log](https://github.com/ml-ichiro/meta-debian/files/14912501/ptest-dbus.log)

## Test summary

* TOTAL: 23
  * PASS: 18
  * FAIL: 3
  * SKIP: 2

I run this ptest 3 times and obtained the same results.

ksasaky committed:
dbus: Respect "ptest" in DISTRO_FEATURES

dbus-test is not installed even if DISTRO_FEATURES contains "ptest",
because dbus recipe doesn't check DISTRO_FEATURES, so fix it.

Signed-off-by: Kazuho Sasaki <sasaki.kazuho@meta.co.jp>

masami256 commented:
@ml-ichiro Is it okay to ignore failed tests? 
AFAIK, following error can be ignored when we run test on not GUI environment.

```
*** Failed to open connection to system bus: Unable to autolaunch a dbus-daemon without a $DISPLAY for X11
```

How about others?

ksasaky commented:
FYI there are some improvements found in poky dunfell as follows:

- [8216015a85](https://github.com/yoctoproject/poky/commit/8216015a85812a225563ac85149ecaf37678d2c6) dbus: fix ptest failure
- [04414b62a3](https://github.com/yoctoproject/poky/commit/04414b62a3e2d0547bcd64a7b79eefb43d839dc2) dbus: disable test-bus
- [bb0b96a586](https://github.com/yoctoproject/poky/commit/bb0b96a586b7dd0246090108b2e8899ee203124b) dbus: fix failing ptest (by disabling as it needs X)

The first commit seems to fix `test-bus` and `test-dbus-daemon`. The second and third commits just skip `test-bus` and `test-names`. `test-names` is the test case which @masami256 pointed out.

I'm not sure the second and third commits are appropriate for meta-debian's policy, but at least the first commit is worth backporting.

masami256 commented:
> FYI there are some improvements found in poky dunfell as follows:
> 
>     * [8216015a85](https://github.com/yoctoproject/poky/commit/8216015a85812a225563ac85149ecaf37678d2c6) dbus: fix ptest failure
> 
>     * [04414b62a3](https://github.com/yoctoproject/poky/commit/04414b62a3e2d0547bcd64a7b79eefb43d839dc2) dbus: disable test-bus
> 
>     * [bb0b96a586](https://github.com/yoctoproject/poky/commit/bb0b96a586b7dd0246090108b2e8899ee203124b) dbus: fix failing ptest (by disabling as it needs X)
> 
> 
> The first commit seems to fix `test-bus` and `test-dbus-daemon`. The second and third commits just skip `test-bus` and `test-names`. `test-names` is the test case which @masami256 pointed out.
> 
> I'm not sure the second and third commits are appropriate for meta-debian's policy, but at least the first commit is worth backporting.

@ksasaky thank you for the analyzing. Fixing ptest failures can be acceptable.
Talking about [04414b62a3](https://github.com/yoctoproject/poky/commit/04414b62a3e2d0547bcd64a7b79eefb43d839dc2) , if PTEST_RUNNER_TIMEOUT environment value prevent timeout, backport is not required.
If ptest required GUI environment, we can treat these tests as known issue on non-GUI environment. It would be nice to these information in commit log.

masami256 commented:
> 
> @ksasaky thank you for the analyzing. Fixing ptest failures can be acceptable. Talking about [04414b62a3](https://github.com/yoctoproject/poky/commit/04414b62a3e2d0547bcd64a7b79eefb43d839dc2) , if PTEST_RUNNER_TIMEOUT environment value prevent timeout, backport is not required. 

Or , if "ptest-runner -t 3600" solves timeout issue backport is not needed.

ksasaky committed:
dbus: Fix ptest failure

Backport the following patch from Poky-dunfell:
- 8216015a85 dbus: fix ptest failure

Note that test-bus testcase usually takes longer than ptest-runner's
default timeout (300s). Please set longer timeout like "ptest-runner -t
3600" if you would like to avoid timeout.

Also note that test-names testcase requires X, so it is still not
supported in non-GUI environment.

ksasaky commented:
I fixed this ptest failure by backporting [8216015a85](https://github.com/yoctoproject/poky/commit/8216015a85812a225563ac85149ecaf37678d2c6) only.

> If ptest required GUI environment, we can treat these tests as known issue on non-GUI environment. It would be nice to these information in commit log.

Thank you for the suggestion!
Actually `test-names` requires X, so we can't run this test case in non-GUI environment.
I leaved some notes about this in the commit log.

> Or , if "ptest-runner -t 3600" solves timeout issue backport is not needed.

OK. I didn't backport [04414b62a3](https://github.com/yoctoproject/poky/commit/04414b62a3e2d0547bcd64a7b79eefb43d839dc2) because `ptest-runner -t 3600` solves timeout issue.


Here is the test log.
[ptest-dbus.log](https://github.com/user-attachments/files/15964879/ptest-dbus.log)

Summary
* PASS: 20
* FAIL: 1
* SKIP: 2

I run this ptest 3 times and obtained the same results.

masami256 commented:
@ksasaky thank you for your work.
